Handle Auth0 Too Many Requests errors

`[ERROR] 2020-07-17 20:06:24,257 utils.py:260 Auth0 Management API profile update failed https://auth.mozilla.auth0.com/api/v2/users/github%7C123456789 : {'user_metadata': {'existsInCIS': False}} : {"statusCode":429,"error":"Too Many Requests"}`

Develop logic to
* Consume the [HTTP headers indicating rate limiting](https://auth0.com/docs/policies/rate-limits#review-http-response-headers)
* Do some kind of backoff (queue up the request in an SQS queue and then process them later?)

--- a/auth0-cis-webhook-consumer.yaml
+++ b/auth0-cis-webhook-consumer.yaml
@@ -229,7 +229,7 @@ Resources:
       # resource is created, creating the LogGroup with no expiration and
       # preventing this resource from creating
       LogGroupName: !Join [ '/', ['/aws/lambda', !Ref 'Auth0CISWebHookConsumerAsyncFunction' ] ]
-      RetentionInDays: 14
+      RetentionInDays: 30
   Auth0CISWebHookConsumerDomainName:
     Type: AWS::ApiGateway::DomainName
     Condition: UseCustomDomainName

--- a/functions/auth0_cis_webhook_consumer/app.py
+++ b/functions/auth0_cis_webhook_consumer/app.py
@@ -58,7 +58,10 @@ def process_api_call(
             # https://github.com/mozilla-iam/cis/blob/73f21ab201b4f242512786dfc8e1707fccf7f3c5/python-modules/cis_notifications/cis_notifications/event.py#L44-L51
             operation = body.get('operation')
             if (user_id is not None and operation is not None
-                    and process_auth0_user(user_id, operation)):
+                    and process_auth0_user(
+                        user_id,
+                        operation,
+                        context.get_remaining_time_in_millis)):
                 return {
                     'headers': {'Content-Type': 'text/html'},
                     'statusCode': 200,
@@ -106,10 +109,15 @@ def lambda_handler(event: LambdaDict, context: LambdaContext) -> LambdaDict:
                 if event['headers'] is not None else {})
             cis_webhook_authorization = headers.get('authorization')
             body = json.loads(event['body'])
-            return process_api_call(event, cis_webhook_authorization, body)
+            return process_api_call(
+                event, context, cis_webhook_authorization, body)
         except json.decoder.JSONDecodeError:
             logger.error('Unable to parse POSTed body : {}'.format(
                 event['body']))
+            return {
+                'headers': {'Content-Type': 'text/html'},
+                'statusCode': 500,
+                'body': 'Error'}
         except Exception as e:
             logger.error(str(e))
             logger.error(traceback.format_exc())

--- a/functions/auth0_cis_webhook_consumer/app.py
+++ b/functions/auth0_cis_webhook_consumer/app.py
@@ -9,6 +9,7 @@ from .utils import (
     verify_token,
     process_auth0_user
 )
+from .lambda_types import LambdaDict, LambdaContext
 
 logger = logging.getLogger()
 if len(logging.getLogger().handlers) == 0:
@@ -24,12 +25,14 @@ CONFIG = Config()
 
 
 def process_api_call(
-        event: dict,
+        event: LambdaDict,
+        context: LambdaContext,
         cis_webhook_authorization: str,
         body: dict) -> dict:
     """Process an API Gateway call depending on the URL path called
 
     :param event: The API Gateway request event
+    :param context: AWS Lambda context object
     :param cis_webhook_authorization: A bearer token from the CIS webhook
            service
     :param body: The parsed body that was POSTed to the API Gateway
@@ -78,7 +81,7 @@ def process_api_call(
             'body': "That path wasn't found"}
 
 
-def lambda_handler(event: dict, context: dict) -> dict:
+def lambda_handler(event: LambdaDict, context: LambdaContext) -> LambdaDict:
     """Handler for all API Gateway requests
 
     :param event: AWS API Gateway input fields for AWS Lambda

--- a/functions/auth0_cis_webhook_consumer/app.py
+++ b/functions/auth0_cis_webhook_consumer/app.py
@@ -89,6 +89,17 @@ def lambda_handler(event: LambdaDict, context: LambdaContext) -> LambdaDict:
     :return: An AWS API Gateway output dictionary for proxy mode
     """
     if event.get('resource') == '/{proxy+}':
+        if event.get('httpMethod') != 'POST':
+            return {
+                'headers': {'Content-Type': 'text/html'},
+                'statusCode': 405,
+                'body': '405 Method Not Allowed'}
+        elif not event.get('body'):
+            logger.debug('Missing POST body')
+            return {
+                'headers': {'Content-Type': 'text/html'},
+                'statusCode': 400,
+                'body': 'Missing POST body'}
         try:
             headers = (
                 {x.lower(): event['headers'][x] for x in event['headers']}

--- a/functions/auth0_cis_webhook_consumer/lambda_types.py
+++ b/functions/auth0_cis_webhook_consumer/lambda_types.py
@@ -1,0 +1,39 @@
+""" Note: this code is used only by the static type checker """
+from typing import Dict, Any
+
+LambdaDict = Dict[str, Any]
+
+
+class LambdaCognitoIdentity(object):
+    cognito_identity_id: str
+    cognito_identity_pool_id: str
+
+
+class LambdaClientContextMobileClient(object):
+    installation_id: str
+    app_title: str
+    app_version_name: str
+    app_version_code: str
+    app_package_name: str
+
+
+class LambdaClientContext(object):
+    client: LambdaClientContextMobileClient
+    custom: LambdaDict
+    env: LambdaDict
+
+
+class LambdaContext(object):
+    function_name: str
+    function_version: str
+    invoked_function_arn: str
+    memory_limit_in_mb: int
+    aws_request_id: str
+    log_group_name: str
+    log_stream_name: str
+    identity: LambdaCognitoIdentity
+    client_context: LambdaClientContext
+
+    @staticmethod
+    def get_remaining_time_in_millis() -> int:
+        return 0

--- a/functions/auth0_cis_webhook_consumer/utils.py
+++ b/functions/auth0_cis_webhook_consumer/utils.py
@@ -20,6 +20,39 @@ def filter_profile(item):
               if key not in ['metadata', 'signature']})
 
 
+def wait_for_ratelimit_reset(reset_time, get_remaining_time_in_millis):
+    """If there's enough execution time remaining before ratelimit resets,
+    sleep until then
+
+    :param reset_time: Seconds since the epoch at which point the ratelimit
+        resets
+    :param get_remaining_time_in_millis: Function that returns the amount of
+        time remaining to complete execution of the AWS Lambda function in
+        milliseconds
+    :return: True if we have enough time to sleep and we've slept, False if not
+    """
+    if reset_time is None:
+        reset_time = 0
+    reset_time = int(reset_time)
+    seconds_until_reset = max(reset_time - time.time(), 0)
+    if get_remaining_time_in_millis() > (seconds_until_reset + 30) * 1000:
+        # We have enough remaining time in execution to sleep
+        logger.debug(
+            'Sleeping for {} seconds until the ratelimit resets and more '
+            'calls are available.'.format(seconds_until_reset))
+        time.sleep(seconds_until_reset)
+        return True
+    else:
+        # We don't have enough time
+        logger.error(
+            'It will be {} seconds until the ratelimit resets and more calls '
+            'are available which exceeds the execution time available to this '
+            'AWS Lambda function of {} milliseconds.'.format(
+                seconds_until_reset,
+                get_remaining_time_in_millis()))
+        return False
+
+
 def verify_token(authorization: str, jwks: dict, issuer: str) -> bool:
     """Verify that bearer token is valid
 
@@ -167,7 +200,10 @@ def hack_user_id(user_id):
         return user_id
 
 
-def process_auth0_user(user_id: str, operation: str) -> bool:
+def process_auth0_user(
+        user_id: str,
+        operation: str,
+        get_remaining_time_in_millis: Callable[[], int]) -> bool:
     """Process the operation on the Auth0 user
 
     Requires Auth0 Management API scopes
@@ -176,6 +212,8 @@ def process_auth0_user(user_id: str, operation: str) -> bool:
 
     :param user_id: The user's user ID
     :param operation: The operation to perform, "create", "update", "delete"
+    :param get_remaining_time_in_millis: Function that returns how much time
+        remains to complete execution
     :return:
     """
     if operation == "create":
@@ -251,18 +289,45 @@ def process_auth0_user(user_id: str, operation: str) -> bool:
                 'Skipping Auth0 update on {} as the user is not in the '
                 'whitelist'.format(user_id))
             return True
-    # https://auth0.com/docs/api/management/v2/#!/Users/patch_users_by_id
-    response = requests.patch(
-        url=url,
-        json=payload,
-        headers=headers)
-    if not response.ok:
-        logger.error(
-            'Auth0 Management API profile update failed {} : {} : {}'.format(
-                url,
-                payload,
-                response.text))
-    else:
-        logger.info('Successfully updated Auth0 user {} : {}'.format(
-            user_id, payload))
-    return response.ok
+
+    update_can_succeed = True
+    global last_auth0_request
+    if 'last_auth0_request' in globals():
+        if last_auth0_request.get('X-RateLimit-Remaining') == 0:
+            # No remaining requests available, sleep until there are
+            if not wait_for_ratelimit_reset(
+                    last_auth0_request.get('X-RateLimit-Reset'),
+                    get_remaining_time_in_millis):
+                update_can_succeed = False
+
+    while update_can_succeed:
+        # https://auth0.com/docs/api/management/v2/#!/Users/patch_users_by_id
+        response = requests.patch(
+            url=url,
+            json=payload,
+            headers=headers)
+        last_auth0_request = response.headers
+        if response.status_code == 429:
+            update_can_succeed = wait_for_ratelimit_reset(
+                response.headers.get('X-RateLimit-Reset'),
+                get_remaining_time_in_millis)
+        else:
+            # We weren't ratelimited
+            if not response.ok:
+                logger.critical(
+                    'Auth0 Management API profile update failed {} : {} : '
+                    '{}'.format(
+                        url,
+                        payload,
+                        response.text))
+                return False
+            break
+    if not update_can_succeed:
+        logger.critical(
+            'Currently ratelimited by Auth0. As a result this update to '
+            'Auth0 can not be sent and will be lost. The user is {} and the '
+            'update is : {}'.format(user_id, payload))
+        return False
+    logger.info('Successfully updated Auth0 user {} : {}'.format(
+        user_id, payload))
+    return True

--- a/functions/auth0_cis_webhook_consumer/utils.py
+++ b/functions/auth0_cis_webhook_consumer/utils.py
@@ -4,7 +4,7 @@ import time
 import requests
 import urllib.parse
 from jose import jwt, exceptions
-from typing import Optional
+from typing import Optional, Callable
 
 from .config import Config
 

